### PR TITLE
Upgrade to TypeScript 4.0

### LINKED ISSUE #8646: Support TypeScript 4.0
TypeScript 4.0 beta has been released!

https://devblogs.microsoft.com/typescript/announcing-typescript-4-0-beta/

We need to support new syntaxes:

- Variadic Tuple Types
- Labeled Tuple Elements
- Short-Circuiting Assignment Operators
  - This syntax is already supported via Babel (https://babeljs.io/docs/en/next/babel-plugin-proposal-logical-assignment-operators.html)

note: We should wait https://github.com/typescript-eslint/typescript-eslint/issues/2260 for default TS parser and https://github.com/babel/babel/pull/11760 for babel-ts parser.

### PR DESCRIPTION
**Pull request by sosukesuzuki**
Closes #8646 

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by sosukesuzuki** [comment]
Oh, we should remove this warning.
スクリーンショット 2020-08-15 16 36 57

File: tests/typescript/tuple/tuple-labeled.ts [hunk 1]
```diff
@@ -14,9 +13,3 @@ type T = [x?: A, y: B];
 type T = [x: A, ...B];
 
 type T = [...B, x: A];
-
-// Invalid label 1
-type T = [x.y: A];
-
-// Invalid label 2
-type T = [x<y>: A];
```
Comment (fisker) [inline tests/typescript/tuple/tuple-labeled.ts:22]: Let's keep these two in `typescript-babel-only`

**Comment by fisker** [comment]
> Oh, we should remove this warning.
> スクリーンショット 2020-08-15 16 36 57

I think we should wait until `typescript-eslint` officially support `v4`

**Comment by fisker** [comment]
There is  an option `warnOnUnsupportedTypeScriptVersion` for this warning? https://github.com/typescript-eslint/typescript-eslint/blob/6f397df42cbcf05c10f304c9bbfdae4803aa0ce2/packages/parser/src/parser.ts#L72

**Comment by sosukesuzuki** [comment]
Set `loggerFn` to false then it works fine.
スクリーンショット 2020-08-19 17 49 56

**Comment by sosukesuzuki** [comment]
According to https://github.com/microsoft/TypeScript/issues/38510, 4.0 will be released on 8/20, so let's wait.

**Review by evilebottnawi — APPROVED** [review]

**Comment by AlCalzone** [comment]
TS 4.0.2 is out now :)

**Comment by petejkim** [comment]
Current version of prettier can't parse the new "Short-Circuiting Assignment Operators" syntax.

e.g.

```typescript
foo ??= "default value"
```

**Comment by fisker** [comment]
> Current version of prettier can't parse the new "Short-Circuiting Assignment Operators" syntax.
> 
> e.g.
> 
> ```ts
> foo ??= "default value"
> ```

@sosukesuzuki Can you check it's tested on `typescript` parser?

**Comment by brodybits** [comment]
There will be type errors if you merge with master, due to recently merged PR #8759:

- duplicate identifiers from these lines: https://github.com/prettier/prettier/blob/cdc250c21ad552667809de2c6f8536d50f1e3f57/src/language-js/types/estree.d.ts#L6-L12
- type issues in `src/language-js/utils.js`

I really hope we don't have to revert anything major.

**Comment by fisker** [comment]
I thought you applied this suggestion https://github.com/prettier/prettier/pull/8759#discussion_r473116995, how to fix it?

**Comment by brodybits** [comment]
You can just remove the workaround hack lines when you merge from master: https://github.com/prettier/prettier/blob/cdc250c21ad552667809de2c6f8536d50f1e3f57/src/language-js/types/estree.d.ts#L6-L12

You will also have to solve the type issues between `src/language-js/utils.js` and TypeScript 4. I think you will want  do this before merging this PR into master. What a mess:)

**Comment by sosukesuzuki** [comment]
> You will also have to solve the type issues between src/language-js/utils.js and TypeScript 4. I think you will want do this before merging this PR into master. What a mess:)

@brodybits How to fix?

**Comment by brodybits** [comment]
> How to fix?

I proposed a quick workaround here: https://github.com/sosukesuzuki/prettier/pull/2

This is using `any` parameter type in 3 places. I recall we discussed elsewhere (@thorn0?) that we should eventually search for `any` type specifiers and replace them with more specific types at some point in the future. But I would definitely not favor blocking TypeScript 4.0 update with this.

**Comment by brodybits** [comment]
> ```ts
> foo ??= "default value"
> ```

seems to be tested with Babel & TypeScript in `tests/js/logical-assignment`

**Comment by sosukesuzuki** [comment]
Thank you. I agree with you. For now, let's merge it and consider a fix after the 2.1 release.

**Comment by fisker** [comment]
Let's merge!

**Comment by sosukesuzuki** [comment]
Should we add changelog for Short-Circuiting Assignment Operators?

**Pull request merged**